src/process.c

```diff
@@ -463,59 +466,450 @@ cc_run_hook(struct oci_cfg_hook* hook, const gchar* state,
-			flags,
-			(GSpawnChildSetupFunc)cc_oci_setup_child,
-			(gpointer)config,
-			&config->state.workload_pid,
-			NULL,
-			NULL,
-			NULL,
-			&err);
+	if (! cc_oci_fd_set_cloexec (child_err_pipe[1])) {
+		g_critical ("failed to set close-exec bit on fd");
+		goto out;
+	}
+
+	pid = config->state.workload_pid = fork ();
+	if (pid < 0) {
+		g_critical ("failed to create child: %s",
+				strerror (errno));
+		goto out;
+	}
+
+	if (! pid) {
+		g_autofree gchar  *netcfg = NULL;
+
+		/* child */
+		pid = config->state.workload_pid = getpid ();
+
+		close (net_cfg_pipe[1]);
+		close (child_err_pipe[0]);
+
+		/* block waiting for parent to send network config.
+		 *
+		 * This arrives in 2 parts: first, a size_t representing
+		 * the size of the ASCII name=value data that follows.
+		 */
+
+		/* determine how many bytes to allocate for the network
+		 * config.
+		 */
+		g_debug ("child reading netcfg length");
+
+		bytes = read (net_cfg_pipe[0], &len, sizeof (len));
+		if (bytes < 0) {
+			goto child_failed;
+		}
+
+		g_debug ("allocating %lu bytes for netcfg",
+				(unsigned long int)len);
+
+		netcfg = g_new0 (gchar, 1+len);
+		if (! netcfg) {
+			goto child_failed;
+		}
+
+		g_debug ("reading netcfg from parent");
+
+		bytes = read (net_cfg_pipe[0], netcfg, len);
+		if (ret < 0) {
+			goto child_failed;
+		}
+
+		close (net_cfg_pipe[0]);
+
+		g_debug ("adding netcfg to config");
+
+		if (! cc_oci_vm_netcfg_from_str (config, netcfg)) {
+			goto child_failed;
+		}
+
+
+		g_debug ("building hypervisor command-line");
 
+		// FIXME: add network config bits to following functions:
+		//
+		// - cc_oci_container_state()
+		// - oci_state()
+		// - cc_oci_update_options()
+
+		ret = cc_oci_vm_args_get (config, &args);
+		if (! (ret && args)) {
+			goto child_failed;
+		}
+
+		// FIXME: add netcfg to state file
+		ret = cc_oci_state_file_create (config, timestamp);
+		if (! ret) {
+			g_critical ("failed to recreate state file");
+			goto out;
+		}
+
+		g_debug ("running command:");
+		for (p = args; p && *p; p++) {
+			g_debug ("arg: '%s'", *p);
+		}
+
+		if (! cc_oci_setup_child (config)) {
+			goto child_failed;
+		}
+
+		if (execvp (args[0], args) < 0) {
+			g_critical ("failed to exec child %s: %s",
+					args[0],
+					strerror (errno));
+			abort ();
+		}
+	}
+
+	/* parent */
+
+	g_debug ("child pid is %u", (unsigned)pid);
+
+	close (net_cfg_pipe[0]);
+	net_cfg_pipe[0] = -1;
+
+	close (child_err_pipe[1]);
+	child_err_pipe[1] = -1;
+
+	/* create state file before hooks run */
+	ret = cc_oci_state_file_create (config, timestamp);
 	if (! ret) {
-		g_critical ("failed to spawn child: %s", err->message);
-		g_error_free (err);
+		g_critical ("failed to create state file");
+		goto out;
+	}
+
+	/* If a hook returns a non-zero exit code, then an error
+	 * including the exit code and the stderr is returned to
+	 * the caller and the container is torn down.
+	 */
+	ret = cc_run_hooks (config->oci.hooks.prestart,
```

Comment:
@mcastelino I'm trying to understand the flow here: It seems you're doing:

1) From the parent process, create the netns
2) Fork the parent process and have the child wait on its networking configuration pipe
3) From the parent process:
    a) Create state file
    b) Run hooks
    c) Build the netcfg (key=value) concatenated network string
    d) Create the network
    d) Send the netcfg to the child
4) From the child process, receive the netcfg blob, parse it and add it to the process arguments
5) From the child process, `execve()`. The child process arguments will contain the networking ones.

Please correct me if I'm misunderstood. If I'm not, I have a few comments/questions:

1) Why can't we still use `g_spawn()` and call `ns_setup()` and `run_hooks()` before spawning ? I'm probably missing something...
2) Having a specific routine for handling the child process would make this new `vm_launch()` more readable.